File: web/js/board.js
import { BOARD_ROWS, BOARD_COLS } from './tetromino.js';
import { absolutePositions, TETROMINO_COLORS } from './tetromino.js';

export function createBoard() {
  return Array.from({ length: BOARD_ROWS }, () => Array(BOARD_COLS).fill(null));
}

export function isInBounds(row, col) {
  return row >= 0 && row < BOARD_ROWS && col >= 0 && col < BOARD_COLS;
}

export function isValid(piece, board) {
  for (const [row, col] of absolutePositions(piece)) {
    if (col < 0 || col >= BOARD_COLS || row >= BOARD_ROWS) return false;
    if (row >= 0 && board[row][col]) return false;
  }
  return true;
}

export function placePiece(board, piece) {
  const next = board.map((row) => [...row]);
  const color = TETROMINO_COLORS[piece.type];

  for (const [row, col] of absolutePositions(piece)) {
    if (isInBounds(row, col)) {
      next[row][col] = color;
    }
  }

  return next;
}

export function clearLines(board) {
  const remaining = board.filter((row) => row.some((cell) => cell === null));
  const cleared = BOARD_ROWS - remaining.length;

  while (remaining.length < BOARD_ROWS) {
    remaining.unshift(Array(BOARD_COLS).fill(null));
  }

  return { board: remaining, cleared };
}

File: NamchonTetris/NamchonTetris/Models/Board.swift
import Foundation

struct Board {
    static let rows = 20
    static let cols = 10

    private(set) var grid: [[Cell?]]

    init() {
        grid = Array(repeating: Array(repeating: nil, count: Board.cols), count: Board.rows)
    }

    func cell(at row: Int, col: Int) -> Cell? {
        guard isInBounds(row: row, col: col) else { return nil }
        return grid[row][col]
    }

    func isInBounds(row: Int, col: Int) -> Bool {
        row >= 0 && row < Board.rows && col >= 0 && col < Board.cols
    }

    mutating func place(piece: ActivePiece) {
        let cell = Cell(color: piece.type.color)
        for (row, col) in piece.absolutePositions() where isInBounds(row: row, col: col) {
            grid[row][col] = cell
        }
    }

    mutating func clearLines() -> Int {
        var cleared = 0
        var newGrid: [[Cell?]] = []

        for row in 0..<Board.rows {
            if grid[row].allSatisfy({ $0 != nil }) {
                cleared += 1
            } else {
                newGrid.append(grid[row])
            }
        }

        while newGrid.count < Board.rows {
            newGrid.insert(Array(repeating: nil, count: Board.cols), at: 0)
        }

        grid = newGrid
        return cleared
    }
}
